### PR TITLE
Make DictionaryCache have better expiry properties

### LINKED ISSUE #13275: State group caches keep full state for too long, bloating the cache size.
We have a complicated set up for caching state look ups to handle both filtered and full state look ups. However, the eviction strategy is sub-optimal: we keep all cached values for a given state group until we stop querying for *any* keys in that state group.

This means that if we pull out and cache the full state for a state group, we will keep the full state in the cache even if we only query one type / state key pair periodically.

Ideally the cache would see that nothing has asked for the full state in a while and evict it.

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #13275

The `DictionaryCache` is a cache designed for caching lookups into mappings, mainly for looking up state groups. It can handle fetching both subsets of the dict, or the entire thing.

Before, this worked by storing a `DictionaryEntry` in an LruCache, updating the stored dictionary whenever we pulled more entries for the DB. The disadvantage of this method is that we'd never expire subsets of the dictionary that we didn't query, so long as that *other* parts of the dictionary were being queried. For example, if we repeatedly fetch the create event of a room, but at some point ask for the full state of the room, we'd never then drop the full state from the cache, even though we only care about the create event.

This leads to severely bloated state group caches.

Instead, this PR splits up the entries it stores in the `LruCache` into two distinct types:
1. Individual entries in the dictionary, under a `(key, dict_key)` (e.g. for state group caches this would be `(state_group, (type, state_key))`). We set the `LruCache` to use the `TreeCache` to enable us to query/invalidate all entries for a given key.
2. The full dict for a given key, under `(key, FullDict.Key)`.

This allow us to expire the individual entries and the full dict separately.

### CODE REVIEW

File: tests/storage/test_state.py [hunk 1]
```diff
@@ -369,7 +369,7 @@ def test_get_state_for_event(self):
         state_dict_ids = cache_entry.value
 
         self.assertEqual(cache_entry.full, False)
-        self.assertEqual(cache_entry.known_absent, {(e1.type, e1.state_key)})
+        self.assertEqual(cache_entry.known_absent, set())
```
Comment (erikjohnston) [inline tests/storage/test_state.py:372]: Note, this was a bit of a dodgy test, as `(e1.type, e1.state_key)` was not *known absent*, as it is in the returned dict.

File: synapse/util/caches/dictionary_cache.py [hunk 2]
```diff
@@ -96,28 +145,107 @@ def get(
...
+
+                # ... therefore the values must be `_PerKeyValue`
+                assert isinstance(dict_value, _PerKeyValue)
+
+                if dict_value.value is _Sentinel.sentinel:
+                    known_absent.add(dict_key)
+                else:
+                    values[dict_key] = dict_value.value
+
+            return DictionaryEntry(False, known_absent, values)
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:184]: move all this to a separate method?

File: synapse/util/caches/dictionary_cache.py [hunk 3]
```diff
@@ -96,28 +145,107 @@ def get(
...
-                )
+
+        if dict_keys is None:
+            # First we check if we have cached the full dict.
+            entry = self.cache.get((key, _FullCacheKey.KEY), _Sentinel.sentinel)
+            if entry is not _Sentinel.sentinel:
+                assert isinstance(entry, dict)
+                return DictionaryEntry(True, set(), entry)
+
+            # If not, check if we have cached any of dict keys.
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:156]:
```suggestion
            # If not, check if we have cached any dict keys at all for this cache key.
```

File: synapse/util/caches/lrucache.py [hunk 4]
```diff
@@ -568,6 +591,42 @@ def cache_get(
...
+        ) -> Union[T, TreeCacheNode]:
+            ...
+
+        @synchronized
+        def cache_get_multi(
+            key: tuple,
+            default: Optional[T] = None,
+            update_metrics: bool = True,
+        ) -> Union[None, T, TreeCacheNode]:
+            """Used only for `TreeCache` to fetch a subtree."""
```
Comment (richvdh) [inline synapse/util/caches/lrucache.py:616]: ... but what does it actually do? What are the arguments, what are the return values?

also it's not used by TreeCache; it is used to expose a `get_multi` method when the backing cache is a `TreeCache`.

File: synapse/util/caches/lrucache.py [hunk 5]
```diff
@@ -555,10 +561,27 @@ def cache_get(
             default: Optional[T] = None,
             callbacks: Collection[Callable[[], None]] = (),
             update_metrics: bool = True,
+            update_last_access: bool = True,
         ) -> Union[None, T, VT]:
+            """Lookup a key in the cache
```
Comment (richvdh) [inline synapse/util/caches/lrucache.py:566]:
```suggestion
            """Look up a key in the cache
```

File: synapse/util/caches/lrucache.py (hunk 4 above)
Comment (richvdh) [reply synapse/util/caches/lrucache.py:616]: ... also, how is it different to a simple `cache_get` with `update_last_access=False`?

File: synapse/util/caches/dictionary_cache.py [hunk 6]
```diff
@@ -96,28 +145,107 @@ def get(
...
+            )
+            if all_entries is _Sentinel.sentinel:
+                return DictionaryEntry(False, set(), {})
+
+            # If there are entries we need to unwrap the returned cache nodes
+            # and `_PerKeyValue` into the `DictionaryEntry`.
+            values = {}
+            known_absent = set()
+            for dict_key, dict_value in iterate_tree_cache_items((), all_entries):
+                dict_key = dict_key[0]
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:169]: why are we doing this? Isn't the key we're interested in the full tuple?

also, it would be nice not to reuse variables with different types.

File: synapse/util/caches/treecache.py [hunk 7]
```diff
@@ -139,3 +139,17 @@ def iterate_tree_cache_entry(d):
...
     else:
         yield d
+
+
+def iterate_tree_cache_items(key, value):
+    """Helper function to iterate over the leaves of a tree, i.e. a dict of that
+    can contain dicts.
+
+    Returns:
+        A generator yielding key/value pairs.
```
Comment (richvdh) [inline synapse/util/caches/treecache.py:149]: It would be good to expand this. Give an example, maybe?

File: tests/storage/test_state.py (hunk 1 above)
Comment (richvdh) [reply tests/storage/test_state.py:372]: so the code was buggy before and is now fixed?

File: synapse/util/caches/dictionary_cache.py [hunk 8]
```diff
@@ -96,28 +145,107 @@ def get(
...
+                dict_value = dict_value.value
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:170]: so dict_value starts off as a `_PerKeyValue`, and we look up its `value` to get a `DV` or sentinel? in which case wtf is going on at line 177 and 179?

so wtf *is* `dict_value` ?

**Review by richvdh — CHANGES_REQUESTED** [review]
I've started reviewing this, but I'm confused and don't want to spend any longer on it right now.

File: tests/storage/test_state.py (hunk 1 above)
Comment (erikjohnston) [reply tests/storage/test_state.py:372]: Yeeees, though the way that it gets used means that the difference doesn't matter.

File: synapse/util/caches/dictionary_cache.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py]: OH. So I had completely forgotten this horror show, as i tried to add sanity by adding type hints to `TreeCache`, except that turned into a nightmare so didn't want to block this on that and then forgot about the initial insanity that prompted me to try and type hints.

What is going on here is that `DictionaryCache` is backed by `LruCache`, which is backed by `TreeCache`. That means that the you end up with values returned from `iterate_tree_cache_items` being `lru_cache._Node[_PerKeyValue]`.

My plan was to make `get_multi` return a generator that correctly unwrapped the returned nodes. Let me do that now

**Comment by erikjohnston** [comment]
LMK if the clarifications aren't enough

File: synapse/util/caches/dictionary_cache.py [hunk 9]
```diff
@@ -96,28 +145,116 @@ def get(
...
+        )
+        if all_entries is _Sentinel.sentinel:
+            return DictionaryEntry(False, set(), {})
+
+        # If there are entries we need to unwrap the returned cache nodes
+        # and `_PerKeyValue` into the `DictionaryEntry`.
+        values = {}
+        known_absent = set()
+        for cache_key, dict_value in all_entries:
+            # The key used for the `TreeCache` is `(key, dict_key)`
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:232]: at this point we don't really care that `self.cache` is backed by a TreeCache.

```suggestion
            # The keys in `self.cache` are `(cache_key, dict_key)`
```

File: synapse/util/caches/dictionary_cache.py [hunk 10]
```diff
@@ -149,20 +286,27 @@ def update(
             # Only update the cache if the caches sequence number matches the
             # number that the cache had before the SELECT was started (SYN-369)
             if fetched_keys is None:
-                self._insert(key, value, set())
+                self.cache[(key, _FullCacheKey.KEY)] = value
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:289]: do we not need to update (or at least invalidate) entries for individual dict keys here as well?

File: synapse/util/caches/dictionary_cache.py [hunk 11]
```diff
@@ -149,20 +286,27 @@ def update(
...
+        self, key: KT, value: Dict[DKT, DV], fetched_keys: Iterable[DKT]
     ) -> None:
-        # We pop and reinsert as we need to tell the cache the size may have
-        # changed
+        """Add the given dictionary values as explicit keys in the cache.
+
+        Args:
+            key
+            value: The dictionary with all the values that we should cache
+            fetched_keys: The full set of keys that were looked up, any keys
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:301]:
```suggestion
            fetched_keys: The full set of dict keys that were looked up. Any keys
```

File: synapse/util/caches/dictionary_cache.py [hunk 12]
```diff
@@ -149,20 +286,27 @@ def update(
...
-        self, key: KT, value: Dict[DKT, DV], known_absent: Iterable[DKT]
+    def _update_subset(
+        self, key: KT, value: Dict[DKT, DV], fetched_keys: Iterable[DKT]
     ) -> None:
-        # We pop and reinsert as we need to tell the cache the size may have
-        # changed
+        """Add the given dictionary values as explicit keys in the cache.
+
+        Args:
+            key
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:299]:
```suggestion
            key: top-level cache key
```

File: synapse/util/caches/dictionary_cache.py [hunk 13]
```diff
@@ -149,20 +286,27 @@ def update(
...
+        for dict_key, dict_value in value.items():
+            self.cache[(key, dict_key)] = _PerKeyValue(dict_value)
 
-        entry: DictionaryEntry = self.cache.pop(key, DictionaryEntry(False, set(), {}))
-        entry.value.update(value)
-        entry.known_absent.update(known_absent)
-        self.cache[key] = entry
+        for dict_key in fetched_keys:
+            if (key, dict_key) in self.cache:
+                continue
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:310]: this doesn't look correct. If `dict_key` was previously in the database but is no longer, then we will be called with an entry for it in `fetched_keys` but not `value`, and we now need to remove it from `self.cache`.

IOW we need:

```suggestion
            if dict_key in value:
                continue
```

(that may or may not be something we *need* to support, but given it's easy, why not?)

File: synapse/storage/databases/state/store.py [hunk 14]
```diff
@@ -202,7 +202,14 @@ def _get_state_for_group_using_cache(
...
         """
-        cache_entry = cache.get(group)
+        # If we are asked explicitly for a subset of keys, we only ask for those
+        # from the cache. This ensures that the `DictionaryCache` can make
+        # better decisions about what to cache and what to expire.
+        dict_keys = None
+        if not state_filter.has_wildcards():
+            dict_keys = state_filter.concrete_types()
+
+        cache_entry = cache.get(group, dict_keys=dict_keys)
```
Comment (richvdh) [inline synapse/storage/databases/state/store.py:212]: ideally we'd separate this to a different PR. It's not directly related to the changes to DictionaryCache.

**Review by richvdh — CHANGES_REQUESTED** [review]
well, I've stared at it all morning, and now understand how it works. A bunch of suggestions, mostly for clarification.

Any chance of testing any of this? We have a `DictCacheTestCase` which we might be able to extend?

File: synapse/util/caches/dictionary_cache.py (hunk 10 above)
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py:312]: I don't think so, if the value in the DB has changed then we should have previously called `invalidate(..)` to delete the stale entries? We could easily drop any existing entries, but I don't think its necessary?

File: synapse/util/caches/lrucache.py [hunk 15]
```diff
@@ -674,6 +748,8 @@ def cache_contains(key: KT) -> bool:
         self.setdefault = cache_set_default
         self.pop = cache_pop
         self.del_multi = cache_del_multi
+        if cache_type is TreeCache:
+            self.get_multi = cache_get_multi
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/caches/lrucache.py:752]: Yeah, though I'm sorta wondering if when we come to doing more intelligent caching we should just rewrite it in rust (because that would definitely solve everything :DDD)

File: synapse/util/caches/treecache.py [hunk 16]
```diff
@@ -139,3 +139,31 @@ def iterate_tree_cache_entry(d):
...
+    The provided key is a tuple that will get prepended to the returned keys.
+
+    Example:
+
+        cache = TreeCache()
+        cache[(1, 1)] = "a"
+        cache[(1, 2)] = "b"
+        cache[(2, 1)] = "c"
+
+        tree_node = cache.get((1,))
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/caches/treecache.py:157]: Yup. Will do.

(I am not a fan of the tree cache anymore after trying to add types to it)

**Comment by erikjohnston** [comment]
> Any chance of testing any of this? We have a `DictCacheTestCase` which we might be able to extend?

As in, testing that the we now can expire the full dicts but keep the specific entries in the cache? Should be doable.

File: synapse/util/caches/dictionary_cache.py [hunk 17]
```diff
@@ -96,28 +145,116 @@ def get(
...
             else:
-                return DictionaryEntry(
-                    entry.full,
-                    entry.known_absent,
-                    {k: entry.value[k] for k in dict_keys if k in entry.value},
-                )
+                values[dict_key] = dict_value.value
 
-        return DictionaryEntry(False, set(), {})
+        return DictionaryEntry(False, known_absent, values)
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py]: (I've ripped it out)

File: synapse/util/caches/dictionary_cache.py [hunk 18]
```diff
@@ -145,21 +170,21 @@ def get(
         Returns:
             DictionaryEntry
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:171]: this doesn't document what happens if `dict_keys` is None, but we don't have the full dict cached. Could it, please?

File: synapse/util/caches/dictionary_cache.py [hunk 19]
```diff
@@ -253,7 +248,11 @@ def invalidate(self, key: KT) -> None:
         # raced with the INSERT don't update the cache (SYN-369)
         self.sequence += 1
 
-        # Del-multi accepts truncated tuples.
+        # We want to drop all information about the dict for the given key, so
+        # we use `del_multi` to delete it all in one go.
+        #
+        # We ignore the type error here `del_mutli` accepts a truncated key
+        # (when the key type is a tuple).
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:255]:
```suggestion
        # We ignore the type error here: `del_multi` accepts a truncated key
        # (when the key type is a tuple).
```

It feels like this is a problem with the type annotation on `del_multi`, but maybe let's gloss over that for now.

File: synapse/util/caches/dictionary_cache.py [hunk 20]
```diff
@@ -117,7 +247,13 @@ def invalidate(self, key: KT) -> None:
...
         # raced with the INSERT don't update the cache (SYN-369)
         self.sequence += 1
-        self.cache.pop(key, None)
+
+        # We want to drop all information about the dict for the given key, so
+        # we use `del_multi` to delete it all in one go.
+        #
+        # We ignore the type error here `del_mutli` accepts a truncated key
+        # (when the key type is a tuple).
+        self.cache.del_multi((key,))  # type: ignore[arg-type]
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:256]: so how did this ever work before?

File: synapse/util/caches/dictionary_cache.py (hunk 10 above)
Comment (richvdh) [reply synapse/util/caches/dictionary_cache.py:288]: fair enough. That seems like an assumption that we should document in the docstring though.

(Also: shouldn't this method be called `set` or `put` or `insert` or something? Maybe a problem for another time though.)

File: synapse/util/caches/lrucache.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/util/caches/lrucache.py:762]: fair I guess. `LruCache` does look like a nice candidate for rewriting in rust.

File: synapse/util/caches/treecache.py [hunk 21]
```diff
@@ -64,6 +64,15 @@ def set(self, key, value) -> None:
         self.size += 1
 
     def get(self, key, default=None):
+        """When `key` is a full key, fetches the value for the given key (if
+        any).
+
+        If `key` is only a partial key (i.e. a truncated tuple) then returns a
+        `TreeCacheNode`, which can be passed to the `iterate_tree_cache_*`
+        functions to iterate over all values in the cache with keys that start
```
Comment (richvdh) [inline synapse/util/caches/treecache.py:72]:
```suggestion
        functions to iterate over all entries in the cache with keys that start
```

**Review by richvdh — COMMENTED** [review]
looks generally good. still a few comments.

File: synapse/util/caches/dictionary_cache.py (hunk 20 above)
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py:256]: It wasn't a `TreeCache` before: it was just using a `DictionaryEntry` as the cache value, and keeping track of whether that was "full" or "partial".

File: synapse/util/caches/dictionary_cache.py (hunk 19 above)
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py:255]: Agreed. I ended up in a bit of a rabbithole when trying to fix the type annotations, but maybe we could do something tactical there.

File: synapse/util/caches/dictionary_cache.py (hunk 20 above)
Comment (richvdh) [reply synapse/util/caches/dictionary_cache.py]: > It wasn't a TreeCache before: 

oh gosh, I'd completely missed that.

File: synapse/util/caches/dictionary_cache.py (hunk 18 above)
Comment (richvdh) [reply synapse/util/caches/dictionary_cache.py:172]: ... please?

I think it returns some fixed shape of DictionaryEntry ?

File: synapse/util/caches/dictionary_cache.py [hunk 22]
```diff
@@ -267,7 +268,10 @@ def update(
         value: Dict[DKT, DV],
         fetched_keys: Optional[Iterable[DKT]] = None,
     ) -> None:
-        """Updates the entry in the cache
+        """Updates the entry in the cache.
+
+        Note: This does *not* invalidate any existing entries for the `key`.
+        When the underlying data is changed `.invalidate(key)` must be called.
```
Comment (richvdh) [inline synapse/util/caches/dictionary_cache.py:274]:
```suggestion
        Note: This does *not* invalidate any existing entries for the `key`.
        In particular, if we add an entry for the cached "full dict" with 
        `fetched_keys=None`, existing entries for individual dict keys are
        not invalidated. Likewise, adding entries for individual keys does
        not invalidate any cached value for the full dict.
        
        In other words: if the underlying data is *changed*, the cache must
        be explicitly invalidated via `.invalidate()`.
```

**Review by richvdh — COMMENTED** [review]
lgtm other than the docs for the behaviour of .get

File: synapse/util/caches/dictionary_cache.py (hunk 18 above)
Comment (erikjohnston) [reply synapse/util/caches/dictionary_cache.py]: Ah, I'd put some docs under the `dict_keys` arg (and there is some in `DictionaryEntry`), but have also added some in the return value too.

**Comment by erikjohnston** [comment]
Woop woop! Thank you :heart:
